pallene/cfg.lua

```diff
@@ -0,0 +1,276 @@
+        Set.insert(self[from].to,new_node_id)
+    end)
+    return {new_node_id}
+end
+-- creates a new edge from 'from' to 'to'
+local function make_edge(self,to,from)
+    if to and from then
+        Set.insert(self[from].to,to)
+        Set.insert(self[to].from,from)
+    end
+end
+
+-- builds the graph and calls new_node and new_edge.
+-- it receives a node to analyze, the current node list and the current loop.
+-- it returns an updated current node list and the current loop.
+
+-- usually, the current_node is a set with a single element, an index, 
+-- representing the last node visited or the size of the cfg.
+-- but when there is a branch meeting, such after ifs or in before loops,
+-- it becomes necessary to join the branches, having a list of current nodes
+
+-- current_loop is modelled as a stack, having a set of indexes as the element.
+
+
+local function build(self,cmd,current_node,current_loop)
+    local tag = cmd._tag
+    if     tag == "ir.Cmd.Nop" then
+        return current_node,current_loop
+    elseif tag == "ir.Cmd.Seq" then
+        local new_nodes,new_loop = current_node,current_loop
+        for _, c in ipairs(cmd.cmds) do
+            new_nodes,new_loop = build(self,c,new_nodes,new_loop)
+        end
+        assert(new_loop == current_loop, 'unmatched loop in seq (impossible)')
+        return new_nodes,new_loop
+    elseif tag == "ir.Cmd.Return" then
+        local new_node = make_node(self,cmd,current_node,current_loop)
+        Set.iforeach(new_node,function(_,node) 
+            make_edge(self,"exit", node)
+
+        end)
+        return new_node,current_loop
+    elseif tag == "ir.Cmd.Break" then
+        local loop = current_loop[#current_loop]
+        assert(loop,'break outside loop (impossible)')
+        Set.iforeach(current_node,function(_,node) 
+            Set.iforeach(loop,function(_,loop_node) 
+                make_edge(self,loop_node, node)
+            end)
+        end)
+        return current_node,current_loop
+    elseif tag == "ir.Cmd.Loop" or tag == "ir.Cmd.For" then
+        -- in the case of For:
+        -- start,limit,step are ir.Value and 
+        -- will be extracted to assignments beforehand
+        -- so it is equal to a simple Loop in this context
+
+        -- put a new empty top in the current_loop stack
+        table.insert(current_loop, {})
+        local new_nodes,new_loop = build(self,cmd.body,current_node,current_loop)
+        assert(new_loop == current_loop,'unmatched loop ending (impossible)')
+        -- at the end of the loop, gets all indexes at the top 
+        -- and points them with what happened before
+        -- to the same with what happened during the loop
+        local top = current_loop[#current_loop]
+
+        -- the condition is the first node of the body
+        -- TODO: remove this hack
+        -- maybe return the first node of the scope, not the only the current
+        -- maybe propose change in the ir
```

Comment:
I don't understand what this hack is doing going on here.

Reply:
The problem is that the program
```
while cond do
   body1
   body2
end
```
generates 
```
seq {cond, body1, body2}
```
so the hack makes an edge between the first and the last elements of the sequence and between the first and any break that happens in between.

I could separate the seq in head and tail and proceed. But I found it weird. In fact I am not seeing any clean solution to that.